### PR TITLE
feat: return ProposalRef on proposals and allow to remove proposal from the store

### PR DESCRIPTION
**Pull request by augustocdias**
This PR adds the possibility to remove a proposal from the store by its reference. In order to do that all functions that adds a proposal to the store, now are returning a proposal reference.

A test was added to ensure that the proposal is indeed being removed and not added in future commits.

This is useful if a client needs to rollback proposal operations.

### CODE REVIEW

**Comment by raphaelrobert** [comment]
This looks like a useful mechanism, but things might change because of #1295. The spec has changed and the proposal reference now also needs more input than what's available in the proposal itself, at least for standalone proposals. I suggest we wait for #1295.

**Comment by duesee** [comment]
Sorry, @augustocdias. #1295 took a bit longer. I hope it will not be too difficult for you to rebase. Your changes are very useful, also for testing.

File: book/src/user_manual/processing.md
Comment (duesee) [inline book/src/user_manual/processing.md:50]:
```suggestion
Operations that add a proposal to the queue, will return its reference. This reference can be used to remove
```

File: openmls/src/group/core_group/proposals.rs
Comment (duesee) [inline openmls/src/group/core_group/proposals.rs:52]:
```suggestion

    pub(crate) fn remove(&mut self, proposal_ref: ProposalRef) -> Option<()> {
```

File: openmls/src/group/mls_group/membership.rs
Comment (duesee) [inline openmls/src/group/mls_group/membership.rs:10]: Could you please use `ciphersuite::hash_ref::ProposalRef` instead?

File: openmls/src/group/mls_group/mod.rs
Comment (duesee) [inline openmls/src/group/mls_group/mod.rs:14]:
```suggestion
    ciphersuite::hash_ref::ProposalRef,
```

File: openmls/src/group/mls_group/mod.rs
Comment (duesee) [inline openmls/src/group/mls_group/mod.rs:428]: I think I would call this `remove` and not clear. "clear" is mostly used as in "clear everything" (see above).

**Review by duesee — COMMENTED** [review]
Thanks! I made some suggestions already but let's wait for another approval.

File: openmls/src/group/core_group/proposals.rs
Comment (duesee) [inline openmls/src/group/core_group/proposals.rs:57]: I'm not sure about this. `swap_remove` changes the order of elements. I think the more safe thing to do would be to just remove the proposal.

File: openmls/src/group/core_group/proposals.rs
Comment (duesee) [reply openmls/src/group/core_group/proposals.rs:52]: (This is only about a newline.)

**Comment by augustocdias** [comment]
> Thanks! I made some suggestions already but let's wait for another approval.

Thanks. I've addressed all comments.

File: openmls/tests/book_code.rs
Comment (raphaelrobert) [inline openmls/tests/book_code.rs:832]: Just a nit, but `remove_proposal_by_ref` sounds confusing since we have an add proposal. Maybe 'roll back'?

File: book/src/user_manual/processing.md
Comment (raphaelrobert) [inline book/src/user_manual/processing.md:51]:
```suggestion
Operations that add a proposal to the proposal store, will return its reference. This reference can be used to remove
a proposal from the proposal store. This can be useful for example to roll back in case of errors.
```

File: openmls/src/group/core_group/proposals.rs
Comment (raphaelrobert) [inline openmls/src/group/core_group/proposals.rs:53]: Can you add some docs?

File: openmls/src/group/mls_group/mod.rs
Comment (raphaelrobert) [inline openmls/src/group/mls_group/mod.rs:435]: The comment says the proposal is returned, but it isn't

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Thanks for the PR, the tests look thorough, just a few comments.

File: openmls/tests/book_code.rs
Comment (augustocdias) [reply openmls/tests/book_code.rs:832]: Would it work if I put the add proposal out of the anchor? Or you prefer to remove the doc completely regarding this?

Comment (augustocdias) [reply openmls/tests/book_code.rs]: I've just got your comment. You meant to rename the anchor, right?

**Comment by raphaelrobert** [comment]
Can you merge main? I can't push to your branch, unfortunately.

**Comment by augustocdias** [comment]
> Can you merge main? I can't push to your branch, unfortunately.

done

**Review by raphaelrobert — APPROVED** [review]

**Pull request merged**
